sys/mips/82371AB.c

```diff
@@ -0,0 +1,141 @@
+  assert(dev->bus == DEV_BUS_ISA);
+
+  if (type != RT_IOPORTS) {
+    device_t *idev = BUS_METHOD_PROVIDER(dev->parent, alloc_resource);
+    return BUS_METHODS(idev->parent)
+      .alloc_resource(idev, type, rid, start, end, size, flags);
+  }
+
+  assert(start >= IO_ICUSIZE);
+
+  rman_t *rman = NULL;
+  intel_isa_state_t *isa = dev->parent->state;
+  rman = &isa->io_rman;
+  bus_space_handle_t bh = isa->io->r_bus_handle;
+
+  resource_t *r =
+    rman_reserve_resource(rman, type, rid, start, end, size, 0, flags);
+
+  if (r == NULL)
+    return NULL;
+
+  /* We shouldn't do that for IRQs, but rn we'll never end up here when
+   * allocating them, because they get dispatched. */
+  r->r_bus_tag = generic_bus_space;
+  r->r_bus_handle = bh + resource_start(r);
+
+  /* IOPorts are always active */
+  if (bus_activate_resource(dev, r)) {
+    resource_release(r);
+    return NULL;
+  }
+
+  return r;
+}
+
+static int intel_isa_activate_resource(device_t *dev, resource_t *r) {
+  /* Neither IRQs in general or IOports on ISA require activation */
+  return 0;
+}
+
+static void intel_isa_deactivate_resource(device_t *dev, resource_t *r) {
+  /* Neither IRQs in general or IOports on ISA require deactivation */
+}
+
+static void intel_isa_release_resource(device_t *dev, resource_t *r) {
+  if (r->r_type != RT_IOPORTS) {
+    device_t *idev = BUS_METHOD_PROVIDER(dev->parent, release_resource);
+    BUS_METHODS(idev->parent).release_resource(idev, r);
+    return;
+  }
+  bus_deactivate_resource(dev, r);
+  resource_release(r);
+}
+
+static int intel_isa_probe(device_t *d) {
+  return d->unit == 0;
+}
+
+static int intel_isa_attach(device_t *isab) {
+  intel_isa_state_t *isa = isab->state;
+  isa->io = device_take_ioports(isab, 0, RF_ACTIVE);
+  rman_init_from_resource(&isa->io_rman, "ISA IO ports", isa->io);
+
+  /* -------------------------------------------------------------
+   * Create child devices of ISA bus.
+   */
+  device_t *dev;
+
+  /* atkbdc keyboard device */
+  dev = device_add_child(isab, 0);
+  dev->bus = DEV_BUS_ISA;
+  device_add_ioports(dev, 0, IO_KBD, IO_KBDSIZE);
+  device_add_irq(dev, 0, 1);
+
+  /* ns16550 uart device */
+  dev = device_add_child(isab, 1);
+  dev->bus = DEV_BUS_ISA;
+  device_add_ioports(dev, 0, IO_COM1, IO_COMSIZE);
+  device_add_irq(dev, 0, 4);
+
+  /* rtc device */
+  dev = device_add_child(isab, 2);
+  dev->bus = DEV_BUS_ISA;
+  device_add_ioports(dev, 0, IO_RTC, IO_RTCSIZE);
+  device_add_irq(dev, 0, 8);
+
+  /* i8254 timer device */
+  dev = device_add_child(isab, 3);
+  dev->bus = DEV_BUS_ISA;
+  device_add_ioports(dev, 0, IO_TIMER1, IO_TMRSIZE);
+  device_add_irq(dev, 0, 0);
+
+  return bus_generic_probe(isab);
+}
+
+static bus_methods_t intel_isa_bus_bus_if = {
+  .intr_setup = NULL,    /* Dispatched */
+  .intr_teardown = NULL, /* Dispatched */
+  .alloc_resource = intel_isa_alloc_resource,
+  .activate_resource = intel_isa_activate_resource,
+  .deactivate_resource = intel_isa_deactivate_resource,
```

Comment:
can it be null? with above (line 65) comment

Reply:
Could you elaborate? Line 65 is currently
```C
static void intel_isa_deactivate_resource(device_t *dev, resource_t *r) {
```
It's `NULL` because these procedure are dispatched to the gt driver. If your concern is that the information about the original caller might get lost due to insufficient method semantics then don't worry - it's actually still there, just hidden within auxiliary `arg` argument.